### ISSUE TITLE
Relation Extractor custom relations

### ISSUE THREAD

**Issue by aoldoni** [stanfordnlp/CoreNLP#359]
Hi All,

Thanks for the great software. I would like to ask you the following please.

When training specific relations to be extracted from custom Entity types, using the Relation Extractor, I noted that the current possible entities are "hard-coded" in some parts, e.g.:
- https://github.com/stanfordnlp/CoreNLP/blob/master/src/edu/stanford/nlp/ie/machinereading/domains/roth/RothEntityExtractor.java#L16
- https://github.com/stanfordnlp/CoreNLP/blob/master/src/edu/stanford/nlp/ie/machinereading/domains/roth/RothCONLL04Reader.java#L64

By modifying these 2 bits, one can re-use the Relation Extractor successfully with custom entities in case its needed, but this requires then a recompilation and an initial troubleshooting as to understand this.

Would you be interested in a pull-request that refactors these hard-coded methods in something that is obtainable from the properties file? E.g.: in the properties file one can indicate a "entitiesPath" option which would then point to a tab separated file with the normalised and not normalised values of these entities as its columns.

If this option is not provided potentially these default hard coded entities can then be used as to maintain the current behaviour.

This would cause potential Relation Extractor workflows with custom entities to be possible without code recompilation.

Please advise.

Again, thanks!

**Comment by J38** [comment]
There has been a lot of interest in making custom relation extraction training available, but I think the path forward is to make it easier to train models for new relations that work with the KBPAnnotator.  I'm going to try to make sure there is clear documentation and any code changes necessary for that for Stanford CoreNLP 3.8.0.

**Comment by aoldoni** [comment]
OK, cool, thanks for the prompt response.

So the understanding is that the input format will be migrated from the Roth CONLL04 format to the KBP format for the training, and at that point this will become flexible.

At this point I have a small [customisation](https://github.com/aoldoni/stanford-corenlp/commit/ae6782266cb40629aefcecbab397da47f808abc3) locally to adjust this and will continue to use such method.

**Comment by J38** [comment]
Yes that would be the plan.  I'm going to start working on this and hopefully it won't take too long.  By the way, if you happen to have any sample training data I could look at I am looking for an example so I can make sure my modifications are working properly.

**Comment by aoldoni** [comment]
Hi @J38 - sorry it took a while for me to respond to you. Would you have an email for me to send the data? Unfortunately I cannot link it to the internet.

**Comment by rpalenik** [comment]
Hello,

Thanks for great Stanford tools!

I would badly need to be able to train RE with custom entities for my project. I am not a professional (java) programmer (I am able to compile from source if proper instructions available) though and do not fully understand how to "change the code" as aoldoni suggested. Is the possibility of training custom relationships with custom entities available in 3.8? If not, how could I use the approach aoldoni suggested? I have available train corpus in original roth format available. Many thanks for reply! I am attaching small sample train file.

[rel_train.txt](https://github.com/stanfordnlp/CoreNLP/files/1066509/rel_train.txt)

**Comment by aoldoni** [comment]
Hi @rpalenik ,
Just regarding "If not, how could I use the approach aoldoni suggested?", please note:
- You could use this fork: https://github.com/aoldoni/stanford-corenlp - it contains the change that I did.
- Re-compile it, instructions here https://github.com/stanfordnlp/CoreNLP#build-instructions
- Then use the new "possibleEntities" attribute in the properties file, as per this example https://github.com/aoldoni/tetre/blob/develop/config/relation.properties#L54

**Comment by rpalenik** [comment]
Hi @aoldoni ,

Many thanks, I would need some more help. I understand I need to:
1. Clone current source file from https://github.com/stanfordnlp/CoreNLP
2. Replace respective files from your repository
3. Compile

However, I got numerous compilation errors. Have I done it wrong? Can you pls. help with the right approach?

thnx.
R.

**Comment by rpalenik** [comment]
Here is the output from the compiler......
[ant_error.txt](https://github.com/stanfordnlp/CoreNLP/files/1071978/ant_error.txt)
